{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img src=\"http://lssds.aura-astronomy.org/winter_school/sites/default/files/sods_atfdds_header01.jpg\" alt=\"La Serena School for Data Science: Applied Tools for Data-driven Sciences\" />\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 1. Introduction:\n",
    "\n",
    "A Dockerfile is a plain text configuration file used to define the steps for building a Docker image. It is a critical component of the Docker ecosystem, allowing developers to automate the process of creating reproducible and consistent container images. Dockerfiles are used to encapsulate the configuration and dependencies of an application or service, making it easier to package, distribute, and deploy software as containers.\n",
    "\n",
    "The Dockerfile typically includes a series of instructions, each on a separate line, that dictate how the image should be constructed. The Docker engine interprets these instructions and builds the image layer by layer. When a new container is created from the image, the file system is constructed based on these layers, making it efficient and fast.\n",
    "\n",
    "Below is an example of a Dockerfile for creating a Jupyter Notebook server using the official Jupyter Docker Stacks image:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FROM jupyter/base-notebook:latest\r\n",
      "\r\n",
      "WORKDIR /home/jovyan/work\r\n",
      "\r\n",
      "COPY . /home/jovyan/work\r\n",
      "\r\n",
      "EXPOSE 8889\r\n",
      "\r\n",
      "RUN pip install pandas matplotlib\r\n",
      "\r\n",
      "CMD [\"jupyter\", \"notebook\", \"--ip=0.0.0.0\", \"--port=8889\",\"--no-browser\"]\r\n"
     ]
    }
   ],
   "source": [
    "!cat Dockerfile"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2. key components of a Dockerfile:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- **2.1 Base Image:**\n",
    "\n",
    "The Dockerfile starts with specifying the base image upon which the custom image will be built. The base image is the starting point for your application's image and contains the operating system and other pre-installed software or libraries.\n",
    "The base image is defined using the FROM instruction. For example:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "FROM jupyter/minimal-notebook"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- **2.2 Environment Configuration:**\n",
    "\n",
    "The Dockerfile can include environment configuration settings that define variables, specify working directories, and set other runtime behaviors.\n",
    "Environment configuration is set using the ENV instruction. For example:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ENV NODE_ENV=production"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- **2.3 Copying Files:**\n",
    "\n",
    "The COPY instruction allows you to copy files and directories from the host system to the image's file system. This step is essential for adding your application code and any necessary resources to the image.\n",
    "For example:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "COPY . /app"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- **2.4 Running Commands:**\n",
    "\n",
    "The RUN instruction is used to execute commands within the image's file system. This is commonly used to install packages, dependencies, and other software needed for your application to run.\n",
    "\n",
    "For example, for a Node.js application, you might run the following command to install Node.js dependencies:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "RUN npm install"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- **2.5 Exposing Ports:**\n",
    "\n",
    "The EXPOSE instruction is used to specify which ports the container should listen on. It does not actually publish the ports but serves as documentation for developers and users of the image.\n",
    "\n",
    "For example, to indicate that the container listens on port 3000, you would add the following line:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "EXPOSE 3000"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- **2.6 Container Execution Command:**\n",
    "\n",
    "The ***CMD*** or ***ENTRYPOINT*** instruction defines the command that should be executed when a container is started from the image. This is usually the main process of your application.\n",
    "\n",
    "For example, for a Node.js application, you might use:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "CMD [\"node\", \"app.js\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Once you have defined your Dockerfile, you can use the ***docker build*** command to build the Docker image based on the instructions provided in the file. This will create a container image that contains all the specified configurations, dependencies, and code necessary to run your application.\n",
    "\n",
    "Dockerfiles are a powerful tool for creating consistent and reproducible container images. They allow developers to define the exact environment needed for their applications, making it easier to share and distribute software across different environments and systems."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 3. Create a step-by-step Dockerfile:\n",
    "\n",
    "Let's create a step-by-step Dockerfile for a Jupyter Notebook server using the official Jupyter Docker Stacks image:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- **3.1: Choose the Base Image**\n",
    "\n",
    "We'll use the official Jupyter Notebook image from Jupyter Docker Stacks as the base image. This image includes the basic Jupyter Notebook installation.\n",
    "Choose the tag for the base image based on your requirements. In this example, we'll use the ***base-notebook*** tag."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "FROM jupyter/base-notebook:latest"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- **3.2: Set the Working Directory**\n",
    "\n",
    "It's a good practice to set the working directory inside the container where your Jupyter Notebook files will be located.\n",
    "We'll set it to ***/home/jovyan/work***, which is the default directory for Jupyter Notebook files in the base image"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "WORKDIR /home/jovyan/work"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- **3.3: Copy Notebook Files**\n",
    "\n",
    "Next, we'll copy the Jupyter Notebook files from the local directory (where the Dockerfile is located) into the container's working directory.\n",
    "Make sure that your Jupyter Notebook files (e.g., ***.ipynb files***) are present in the same directory as the Dockerfile."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "COPY . /home/jovyan/work"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- **3.4: Expose Port**\n",
    "\n",
    "The Jupyter Notebook runs on port 8888 by default but there are cases where this port may be used, so we will change it to 8889. To access it from the host machine, we need to expose this port from the container."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "EXPOSE 8889"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- **3.5: Install Additional Packages (Optional)**\n",
    "\n",
    "If your Jupyter Notebooks require additional Python packages or libraries, you can install them using pip or conda.\n",
    "In this example, we'll use pip to install pandas and matplotlib."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "RUN pip install pandas matplotlib"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- **3.6: Start Jupyter Notebook Server**\n",
    "\n",
    "Finally, we'll add the command to start the Jupyter Notebook server when the container is run.\n",
    "The --ip=0.0.0.0 option allows access to the server from all IP addresses, and --no-browser prevents Jupyter from trying to open a browser inside the container."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "CMD [\"jupyter\", \"notebook\", \"--ip=0.0.0.0\", \"--port=8889\",\"--no-browser\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The complete Dockerfile will look like this:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FROM jupyter/base-notebook:latest\r\n",
      "\r\n",
      "WORKDIR /home/jovyan/work\r\n",
      "\r\n",
      "COPY . /home/jovyan/work\r\n",
      "\r\n",
      "EXPOSE 8889\r\n",
      "\r\n",
      "RUN pip install pandas matplotlib\r\n",
      "\r\n",
      "CMD [\"jupyter\", \"notebook\", \"--ip=0.0.0.0\", \"--port=8889\",\"--no-browser\"]\r\n"
     ]
    }
   ],
   "source": [
    "!cat Dockerfile"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To build the Docker image, save the above code into a file named Dockerfile in the same directory as your Jupyter Notebook files. Then, open a terminal or command prompt, navigate to the directory containing the Dockerfile, and run the following command:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[1A\u001b[1B\u001b[0G\u001b[?25l[+] Building 0.0s (0/1)                                                         \n",
      "\u001b[?25h\u001b[1A\u001b[0G\u001b[?25l[+] Building 0.1s (2/3)                                                         \n",
      "\u001b[34m => [internal] load build definition from Dockerfile                       0.0s\n",
      "\u001b[0m\u001b[34m => => transferring dockerfile: 37B                                        0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load .dockerignore                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 2B                                            0.0s\n",
      "\u001b[0m => [internal] load metadata for docker.io/jupyter/base-notebook:latest    0.0s\n",
      "\u001b[?25h\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[0G\u001b[?25l[+] Building 0.3s (7/8)                                                         \n",
      "\u001b[34m => [internal] load build definition from Dockerfile                       0.0s\n",
      "\u001b[0m\u001b[34m => => transferring dockerfile: 37B                                        0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load .dockerignore                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 2B                                            0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load metadata for docker.io/jupyter/base-notebook:latest    0.0s\n",
      "\u001b[0m\u001b[34m => [1/4] FROM docker.io/jupyter/base-notebook:latest                      0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load build context                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 558.53kB                                      0.0s\n",
      "\u001b[0m\u001b[34m => CACHED [2/4] WORKDIR /home/jovyan/work                                 0.0s\n",
      "\u001b[0m\u001b[34m => [3/4] COPY . /home/jovyan/work                                         0.1s\n",
      "\u001b[0m => [4/4] RUN pip install pandas matplotlib                                0.0s\n",
      "\u001b[?25h\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[0G\u001b[?25l[+] Building 0.4s (7/8)                                                         \n",
      "\u001b[34m => [internal] load build definition from Dockerfile                       0.0s\n",
      "\u001b[0m\u001b[34m => => transferring dockerfile: 37B                                        0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load .dockerignore                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 2B                                            0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load metadata for docker.io/jupyter/base-notebook:latest    0.0s\n",
      "\u001b[0m\u001b[34m => [1/4] FROM docker.io/jupyter/base-notebook:latest                      0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load build context                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 558.53kB                                      0.0s\n",
      "\u001b[0m\u001b[34m => CACHED [2/4] WORKDIR /home/jovyan/work                                 0.0s\n",
      "\u001b[0m\u001b[34m => [3/4] COPY . /home/jovyan/work                                         0.1s\n",
      "\u001b[0m => [4/4] RUN pip install pandas matplotlib                                0.2s\n",
      "\u001b[?25h\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[0G\u001b[?25l[+] Building 0.6s (7/8)                                                         \n",
      "\u001b[34m => [internal] load build definition from Dockerfile                       0.0s\n",
      "\u001b[0m\u001b[34m => => transferring dockerfile: 37B                                        0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load .dockerignore                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 2B                                            0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load metadata for docker.io/jupyter/base-notebook:latest    0.0s\n",
      "\u001b[0m\u001b[34m => [1/4] FROM docker.io/jupyter/base-notebook:latest                      0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load build context                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 558.53kB                                      0.0s\n",
      "\u001b[0m\u001b[34m => CACHED [2/4] WORKDIR /home/jovyan/work                                 0.0s\n",
      "\u001b[0m\u001b[34m => [3/4] COPY . /home/jovyan/work                                         0.1s\n",
      "\u001b[0m => [4/4] RUN pip install pandas matplotlib                                0.4s\n",
      "\u001b[?25h\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[0G\u001b[?25l[+] Building 0.7s (7/8)                                                         \n",
      "\u001b[34m => [internal] load build definition from Dockerfile                       0.0s\n",
      "\u001b[0m\u001b[34m => => transferring dockerfile: 37B                                        0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load .dockerignore                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 2B                                            0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load metadata for docker.io/jupyter/base-notebook:latest    0.0s\n",
      "\u001b[0m\u001b[34m => [1/4] FROM docker.io/jupyter/base-notebook:latest                      0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load build context                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 558.53kB                                      0.0s\n",
      "\u001b[0m\u001b[34m => CACHED [2/4] WORKDIR /home/jovyan/work                                 0.0s\n",
      "\u001b[0m\u001b[34m => [3/4] COPY . /home/jovyan/work                                         0.1s\n",
      "\u001b[0m => [4/4] RUN pip install pandas matplotlib                                0.5s\n",
      "\u001b[?25h\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[0G\u001b[?25l[+] Building 0.9s (7/8)                                                         \n",
      "\u001b[34m => [internal] load build definition from Dockerfile                       0.0s\n",
      "\u001b[0m\u001b[34m => => transferring dockerfile: 37B                                        0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load .dockerignore                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 2B                                            0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load metadata for docker.io/jupyter/base-notebook:latest    0.0s\n",
      "\u001b[0m\u001b[34m => [1/4] FROM docker.io/jupyter/base-notebook:latest                      0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load build context                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 558.53kB                                      0.0s\n",
      "\u001b[0m\u001b[34m => CACHED [2/4] WORKDIR /home/jovyan/work                                 0.0s\n",
      "\u001b[0m\u001b[34m => [3/4] COPY . /home/jovyan/work                                         0.1s\n",
      "\u001b[0m => [4/4] RUN pip install pandas matplotlib                                0.7s\n",
      "\u001b[?25h\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[0G\u001b[?25l[+] Building 1.0s (7/8)                                                         \n",
      "\u001b[34m => [internal] load build definition from Dockerfile                       0.0s\n",
      "\u001b[0m\u001b[34m => => transferring dockerfile: 37B                                        0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load .dockerignore                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 2B                                            0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load metadata for docker.io/jupyter/base-notebook:latest    0.0s\n",
      "\u001b[0m\u001b[34m => [1/4] FROM docker.io/jupyter/base-notebook:latest                      0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load build context                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 558.53kB                                      0.0s\n",
      "\u001b[0m\u001b[34m => CACHED [2/4] WORKDIR /home/jovyan/work                                 0.0s\n",
      "\u001b[0m\u001b[34m => [3/4] COPY . /home/jovyan/work                                         0.1s\n",
      "\u001b[0m => [4/4] RUN pip install pandas matplotlib                                0.8s\n",
      "\u001b[2m => => # WARNING: Retrying (Retry(total=4, connect=None, read=None, redirect=No\n",
      "\u001b[0m\u001b[2m => => # ne, status=None)) after connection broken by 'NewConnectionError('<pip\n",
      "\u001b[0m\u001b[2m => => # ._vendor.urllib3.connection.HTTPSConnection object at 0x7f4508444210>:\n",
      "\u001b[0m\u001b[2m => => #  Failed to establish a new connection: [Errno -2] Name or service not \n",
      "\u001b[0m\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[?25h\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[0G\u001b[?25l[+] Building 1.2s (7/8)                                                         \n",
      "\u001b[34m => [internal] load build definition from Dockerfile                       0.0s\n",
      "\u001b[0m\u001b[34m => => transferring dockerfile: 37B                                        0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load .dockerignore                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 2B                                            0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load metadata for docker.io/jupyter/base-notebook:latest    0.0s\n",
      "\u001b[0m\u001b[34m => [1/4] FROM docker.io/jupyter/base-notebook:latest                      0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load build context                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 558.53kB                                      0.0s\n",
      "\u001b[0m\u001b[34m => CACHED [2/4] WORKDIR /home/jovyan/work                                 0.0s\n",
      "\u001b[0m\u001b[34m => [3/4] COPY . /home/jovyan/work                                         0.1s\n",
      "\u001b[0m => [4/4] RUN pip install pandas matplotlib                                1.0s\n",
      "\u001b[2m => => # WARNING: Retrying (Retry(total=4, connect=None, read=None, redirect=No\n",
      "\u001b[0m\u001b[2m => => # ne, status=None)) after connection broken by 'NewConnectionError('<pip\n",
      "\u001b[0m\u001b[2m => => # ._vendor.urllib3.connection.HTTPSConnection object at 0x7f4508444210>:\n",
      "\u001b[0m\u001b[2m => => #  Failed to establish a new connection: [Errno -2] Name or service not \n",
      "\u001b[0m\u001b[2m => => # known')': /simple/pandas/                                             \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0m\u001b[?25h\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[0G\u001b[?25l[+] Building 1.4s (7/8)                                                         \n",
      "\u001b[34m => [internal] load build definition from Dockerfile                       0.0s\n",
      "\u001b[0m\u001b[34m => => transferring dockerfile: 37B                                        0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load .dockerignore                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 2B                                            0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load metadata for docker.io/jupyter/base-notebook:latest    0.0s\n",
      "\u001b[0m\u001b[34m => [1/4] FROM docker.io/jupyter/base-notebook:latest                      0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load build context                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 558.53kB                                      0.0s\n",
      "\u001b[0m\u001b[34m => CACHED [2/4] WORKDIR /home/jovyan/work                                 0.0s\n",
      "\u001b[0m\u001b[34m => [3/4] COPY . /home/jovyan/work                                         0.1s\n",
      "\u001b[0m => [4/4] RUN pip install pandas matplotlib                                1.1s\n",
      "\u001b[2m => => # WARNING: Retrying (Retry(total=4, connect=None, read=None, redirect=No\n",
      "\u001b[0m\u001b[2m => => # ne, status=None)) after connection broken by 'NewConnectionError('<pip\n",
      "\u001b[0m\u001b[2m => => # ._vendor.urllib3.connection.HTTPSConnection object at 0x7f4508444210>:\n",
      "\u001b[0m\u001b[2m => => #  Failed to establish a new connection: [Errno -2] Name or service not \n",
      "\u001b[0m\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[?25h\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[0G\u001b[?25l[+] Building 1.5s (7/8)                                                         \n",
      "\u001b[34m => [internal] load build definition from Dockerfile                       0.0s\n",
      "\u001b[0m\u001b[34m => => transferring dockerfile: 37B                                        0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load .dockerignore                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 2B                                            0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load metadata for docker.io/jupyter/base-notebook:latest    0.0s\n",
      "\u001b[0m\u001b[34m => [1/4] FROM docker.io/jupyter/base-notebook:latest                      0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load build context                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 558.53kB                                      0.0s\n",
      "\u001b[0m\u001b[34m => CACHED [2/4] WORKDIR /home/jovyan/work                                 0.0s\n",
      "\u001b[0m\u001b[34m => [3/4] COPY . /home/jovyan/work                                         0.1s\n",
      "\u001b[0m => [4/4] RUN pip install pandas matplotlib                                1.3s\n",
      "\u001b[2m => => # WARNING: Retrying (Retry(total=4, connect=None, read=None, redirect=No\n",
      "\u001b[0m\u001b[2m => => # ne, status=None)) after connection broken by 'NewConnectionError('<pip\n",
      "\u001b[0m\u001b[2m => => # ._vendor.urllib3.connection.HTTPSConnection object at 0x7f4508444210>:\n",
      "\u001b[0m\u001b[2m => => #  Failed to establish a new connection: [Errno -2] Name or service not \n",
      "\u001b[0m\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[?25h\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[0G\u001b[?25l[+] Building 1.7s (7/8)                                                         \n",
      "\u001b[34m => [internal] load build definition from Dockerfile                       0.0s\n",
      "\u001b[0m\u001b[34m => => transferring dockerfile: 37B                                        0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load .dockerignore                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 2B                                            0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load metadata for docker.io/jupyter/base-notebook:latest    0.0s\n",
      "\u001b[0m\u001b[34m => [1/4] FROM docker.io/jupyter/base-notebook:latest                      0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load build context                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 558.53kB                                      0.0s\n",
      "\u001b[0m\u001b[34m => CACHED [2/4] WORKDIR /home/jovyan/work                                 0.0s\n",
      "\u001b[0m\u001b[34m => [3/4] COPY . /home/jovyan/work                                         0.1s\n",
      "\u001b[0m => [4/4] RUN pip install pandas matplotlib                                1.4s\n",
      "\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[2m => => # WARNING: Retrying (Retry(total=3, connect=None, read=None, redirect=No\n",
      "\u001b[0m\u001b[2m => => # ne, status=None)) after connection broken by 'NewConnectionError('<pip\n",
      "\u001b[0m\u001b[2m => => # ._vendor.urllib3.connection.HTTPSConnection object at 0x7f4508444d90>:\n",
      "\u001b[0m\u001b[2m => => #  Failed to establish a new connection: [Errno -2] Name or service not \n",
      "\u001b[0m\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[?25h\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[0G\u001b[?25l[+] Building 1.8s (7/8)                                                         \n",
      "\u001b[34m => [internal] load build definition from Dockerfile                       0.0s\n",
      "\u001b[0m\u001b[34m => => transferring dockerfile: 37B                                        0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load .dockerignore                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 2B                                            0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load metadata for docker.io/jupyter/base-notebook:latest    0.0s\n",
      "\u001b[0m\u001b[34m => [1/4] FROM docker.io/jupyter/base-notebook:latest                      0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load build context                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 558.53kB                                      0.0s\n",
      "\u001b[0m\u001b[34m => CACHED [2/4] WORKDIR /home/jovyan/work                                 0.0s\n",
      "\u001b[0m\u001b[34m => [3/4] COPY . /home/jovyan/work                                         0.1s\n",
      "\u001b[0m => [4/4] RUN pip install pandas matplotlib                                1.6s\n",
      "\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[2m => => # WARNING: Retrying (Retry(total=3, connect=None, read=None, redirect=No\n",
      "\u001b[0m\u001b[2m => => # ne, status=None)) after connection broken by 'NewConnectionError('<pip\n",
      "\u001b[0m\u001b[2m => => # ._vendor.urllib3.connection.HTTPSConnection object at 0x7f4508444d90>:\n",
      "\u001b[0m\u001b[2m => => #  Failed to establish a new connection: [Errno -2] Name or service not \n",
      "\u001b[0m\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[?25h\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[0G\u001b[?25l[+] Building 2.0s (7/8)                                                         \n",
      "\u001b[34m => [internal] load build definition from Dockerfile                       0.0s\n",
      "\u001b[0m\u001b[34m => => transferring dockerfile: 37B                                        0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load .dockerignore                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 2B                                            0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load metadata for docker.io/jupyter/base-notebook:latest    0.0s\n",
      "\u001b[0m\u001b[34m => [1/4] FROM docker.io/jupyter/base-notebook:latest                      0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load build context                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 558.53kB                                      0.0s\n",
      "\u001b[0m\u001b[34m => CACHED [2/4] WORKDIR /home/jovyan/work                                 0.0s\n",
      "\u001b[0m\u001b[34m => [3/4] COPY . /home/jovyan/work                                         0.1s\n",
      "\u001b[0m => [4/4] RUN pip install pandas matplotlib                                1.7s\n",
      "\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[2m => => # WARNING: Retrying (Retry(total=3, connect=None, read=None, redirect=No\n",
      "\u001b[0m\u001b[2m => => # ne, status=None)) after connection broken by 'NewConnectionError('<pip\n",
      "\u001b[0m\u001b[2m => => # ._vendor.urllib3.connection.HTTPSConnection object at 0x7f4508444d90>:\n",
      "\u001b[0m\u001b[2m => => #  Failed to establish a new connection: [Errno -2] Name or service not \n",
      "\u001b[0m\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[?25h\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[0G\u001b[?25l[+] Building 2.1s (7/8)                                                         \n",
      "\u001b[34m => [internal] load build definition from Dockerfile                       0.0s\n",
      "\u001b[0m\u001b[34m => => transferring dockerfile: 37B                                        0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load .dockerignore                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 2B                                            0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load metadata for docker.io/jupyter/base-notebook:latest    0.0s\n",
      "\u001b[0m\u001b[34m => [1/4] FROM docker.io/jupyter/base-notebook:latest                      0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load build context                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 558.53kB                                      0.0s\n",
      "\u001b[0m\u001b[34m => CACHED [2/4] WORKDIR /home/jovyan/work                                 0.0s\n",
      "\u001b[0m\u001b[34m => [3/4] COPY . /home/jovyan/work                                         0.1s\n",
      "\u001b[0m => [4/4] RUN pip install pandas matplotlib                                1.9s\n",
      "\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[2m => => # WARNING: Retrying (Retry(total=3, connect=None, read=None, redirect=No\n",
      "\u001b[0m\u001b[2m => => # ne, status=None)) after connection broken by 'NewConnectionError('<pip\n",
      "\u001b[0m\u001b[2m => => # ._vendor.urllib3.connection.HTTPSConnection object at 0x7f4508444d90>:\n",
      "\u001b[0m\u001b[2m => => #  Failed to establish a new connection: [Errno -2] Name or service not \n",
      "\u001b[0m\u001b[2m => => # known')': /simple/pandas/                                             \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0m\u001b[?25h\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[0G\u001b[?25l[+] Building 2.3s (7/8)                                                         \n",
      "\u001b[34m => [internal] load build definition from Dockerfile                       0.0s\n",
      "\u001b[0m\u001b[34m => => transferring dockerfile: 37B                                        0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load .dockerignore                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 2B                                            0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load metadata for docker.io/jupyter/base-notebook:latest    0.0s\n",
      "\u001b[0m\u001b[34m => [1/4] FROM docker.io/jupyter/base-notebook:latest                      0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load build context                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 558.53kB                                      0.0s\n",
      "\u001b[0m\u001b[34m => CACHED [2/4] WORKDIR /home/jovyan/work                                 0.0s\n",
      "\u001b[0m\u001b[34m => [3/4] COPY . /home/jovyan/work                                         0.1s\n",
      "\u001b[0m => [4/4] RUN pip install pandas matplotlib                                2.0s\n",
      "\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[2m => => # WARNING: Retrying (Retry(total=3, connect=None, read=None, redirect=No\n",
      "\u001b[0m\u001b[2m => => # ne, status=None)) after connection broken by 'NewConnectionError('<pip\n",
      "\u001b[0m\u001b[2m => => # ._vendor.urllib3.connection.HTTPSConnection object at 0x7f4508444d90>:\n",
      "\u001b[0m\u001b[2m => => #  Failed to establish a new connection: [Errno -2] Name or service not \n",
      "\u001b[0m\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[?25h\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[0G\u001b[?25l[+] Building 2.4s (7/8)                                                         \n",
      "\u001b[34m => [internal] load build definition from Dockerfile                       0.0s\n",
      "\u001b[0m\u001b[34m => => transferring dockerfile: 37B                                        0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load .dockerignore                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 2B                                            0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load metadata for docker.io/jupyter/base-notebook:latest    0.0s\n",
      "\u001b[0m\u001b[34m => [1/4] FROM docker.io/jupyter/base-notebook:latest                      0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load build context                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 558.53kB                                      0.0s\n",
      "\u001b[0m\u001b[34m => CACHED [2/4] WORKDIR /home/jovyan/work                                 0.0s\n",
      "\u001b[0m\u001b[34m => [3/4] COPY . /home/jovyan/work                                         0.1s\n",
      "\u001b[0m => [4/4] RUN pip install pandas matplotlib                                2.2s\n",
      "\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[2m => => # WARNING: Retrying (Retry(total=3, connect=None, read=None, redirect=No\n",
      "\u001b[0m\u001b[2m => => # ne, status=None)) after connection broken by 'NewConnectionError('<pip\n",
      "\u001b[0m\u001b[2m => => # ._vendor.urllib3.connection.HTTPSConnection object at 0x7f4508444d90>:\n",
      "\u001b[0m\u001b[2m => => #  Failed to establish a new connection: [Errno -2] Name or service not \n",
      "\u001b[0m\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[?25h\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[0G\u001b[?25l[+] Building 2.5s (7/8)                                                         \n",
      "\u001b[34m => [internal] load build definition from Dockerfile                       0.0s\n",
      "\u001b[0m\u001b[34m => => transferring dockerfile: 37B                                        0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load .dockerignore                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 2B                                            0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load metadata for docker.io/jupyter/base-notebook:latest    0.0s\n",
      "\u001b[0m\u001b[34m => [1/4] FROM docker.io/jupyter/base-notebook:latest                      0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load build context                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 558.53kB                                      0.0s\n",
      "\u001b[0m\u001b[34m => CACHED [2/4] WORKDIR /home/jovyan/work                                 0.0s\n",
      "\u001b[0m\u001b[34m => [3/4] COPY . /home/jovyan/work                                         0.1s\n",
      "\u001b[0m => [4/4] RUN pip install pandas matplotlib                                2.3s\n",
      "\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[2m => => # WARNING: Retrying (Retry(total=2, connect=None, read=None, redirect=No\n",
      "\u001b[0m\u001b[2m => => # ne, status=None)) after connection broken by 'NewConnectionError('<pip\n",
      "\u001b[0m\u001b[2m => => # ._vendor.urllib3.connection.HTTPSConnection object at 0x7f4508445450>:\n",
      "\u001b[0m\u001b[2m => => #  Failed to establish a new connection: [Errno -2] Name or service not \n",
      "\u001b[0m\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[?25h\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[0G\u001b[?25l[+] Building 2.7s (7/8)                                                         \n",
      "\u001b[34m => [internal] load build definition from Dockerfile                       0.0s\n",
      "\u001b[0m\u001b[34m => => transferring dockerfile: 37B                                        0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load .dockerignore                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 2B                                            0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load metadata for docker.io/jupyter/base-notebook:latest    0.0s\n",
      "\u001b[0m\u001b[34m => [1/4] FROM docker.io/jupyter/base-notebook:latest                      0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load build context                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 558.53kB                                      0.0s\n",
      "\u001b[0m\u001b[34m => CACHED [2/4] WORKDIR /home/jovyan/work                                 0.0s\n",
      "\u001b[0m\u001b[34m => [3/4] COPY . /home/jovyan/work                                         0.1s\n",
      "\u001b[0m => [4/4] RUN pip install pandas matplotlib                                2.4s\n",
      "\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[2m => => # WARNING: Retrying (Retry(total=2, connect=None, read=None, redirect=No\n",
      "\u001b[0m\u001b[2m => => # ne, status=None)) after connection broken by 'NewConnectionError('<pip\n",
      "\u001b[0m\u001b[2m => => # ._vendor.urllib3.connection.HTTPSConnection object at 0x7f4508445450>:\n",
      "\u001b[0m\u001b[2m => => #  Failed to establish a new connection: [Errno -2] Name or service not \n",
      "\u001b[0m\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[?25h\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[0G\u001b[?25l[+] Building 2.8s (7/8)                                                         \n",
      "\u001b[34m => [internal] load build definition from Dockerfile                       0.0s\n",
      "\u001b[0m\u001b[34m => => transferring dockerfile: 37B                                        0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load .dockerignore                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 2B                                            0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load metadata for docker.io/jupyter/base-notebook:latest    0.0s\n",
      "\u001b[0m\u001b[34m => [1/4] FROM docker.io/jupyter/base-notebook:latest                      0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load build context                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 558.53kB                                      0.0s\n",
      "\u001b[0m\u001b[34m => CACHED [2/4] WORKDIR /home/jovyan/work                                 0.0s\n",
      "\u001b[0m\u001b[34m => [3/4] COPY . /home/jovyan/work                                         0.1s\n",
      "\u001b[0m => [4/4] RUN pip install pandas matplotlib                                2.6s\n",
      "\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[2m => => # WARNING: Retrying (Retry(total=2, connect=None, read=None, redirect=No\n",
      "\u001b[0m\u001b[2m => => # ne, status=None)) after connection broken by 'NewConnectionError('<pip\n",
      "\u001b[0m\u001b[2m => => # ._vendor.urllib3.connection.HTTPSConnection object at 0x7f4508445450>:\n",
      "\u001b[0m\u001b[2m => => #  Failed to establish a new connection: [Errno -2] Name or service not \n",
      "\u001b[0m\u001b[2m => => # known')': /simple/pandas/                                             \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0m\u001b[?25h\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[0G\u001b[?25l[+] Building 3.0s (7/8)                                                         \n",
      "\u001b[34m => [internal] load build definition from Dockerfile                       0.0s\n",
      "\u001b[0m\u001b[34m => => transferring dockerfile: 37B                                        0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load .dockerignore                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 2B                                            0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load metadata for docker.io/jupyter/base-notebook:latest    0.0s\n",
      "\u001b[0m\u001b[34m => [1/4] FROM docker.io/jupyter/base-notebook:latest                      0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load build context                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 558.53kB                                      0.0s\n",
      "\u001b[0m\u001b[34m => CACHED [2/4] WORKDIR /home/jovyan/work                                 0.0s\n",
      "\u001b[0m\u001b[34m => [3/4] COPY . /home/jovyan/work                                         0.1s\n",
      "\u001b[0m => [4/4] RUN pip install pandas matplotlib                                2.7s\n",
      "\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[2m => => # WARNING: Retrying (Retry(total=2, connect=None, read=None, redirect=No\n",
      "\u001b[0m\u001b[2m => => # ne, status=None)) after connection broken by 'NewConnectionError('<pip\n",
      "\u001b[0m\u001b[2m => => # ._vendor.urllib3.connection.HTTPSConnection object at 0x7f4508445450>:\n",
      "\u001b[0m\u001b[2m => => #  Failed to establish a new connection: [Errno -2] Name or service not \n",
      "\u001b[0m\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[?25h\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[0G\u001b[?25l[+] Building 3.1s (7/8)                                                         \n",
      "\u001b[34m => [internal] load build definition from Dockerfile                       0.0s\n",
      "\u001b[0m\u001b[34m => => transferring dockerfile: 37B                                        0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load .dockerignore                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 2B                                            0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load metadata for docker.io/jupyter/base-notebook:latest    0.0s\n",
      "\u001b[0m\u001b[34m => [1/4] FROM docker.io/jupyter/base-notebook:latest                      0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load build context                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 558.53kB                                      0.0s\n",
      "\u001b[0m\u001b[34m => CACHED [2/4] WORKDIR /home/jovyan/work                                 0.0s\n",
      "\u001b[0m\u001b[34m => [3/4] COPY . /home/jovyan/work                                         0.1s\n",
      "\u001b[0m => [4/4] RUN pip install pandas matplotlib                                2.9s\n",
      "\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[2m => => # WARNING: Retrying (Retry(total=2, connect=None, read=None, redirect=No\n",
      "\u001b[0m\u001b[2m => => # ne, status=None)) after connection broken by 'NewConnectionError('<pip\n",
      "\u001b[0m\u001b[2m => => # ._vendor.urllib3.connection.HTTPSConnection object at 0x7f4508445450>:\n",
      "\u001b[0m\u001b[2m => => #  Failed to establish a new connection: [Errno -2] Name or service not \n",
      "\u001b[0m\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[?25h\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[0G\u001b[?25l[+] Building 3.3s (7/8)                                                         \n",
      "\u001b[34m => [internal] load build definition from Dockerfile                       0.0s\n",
      "\u001b[0m\u001b[34m => => transferring dockerfile: 37B                                        0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load .dockerignore                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 2B                                            0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load metadata for docker.io/jupyter/base-notebook:latest    0.0s\n",
      "\u001b[0m\u001b[34m => [1/4] FROM docker.io/jupyter/base-notebook:latest                      0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load build context                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 558.53kB                                      0.0s\n",
      "\u001b[0m\u001b[34m => CACHED [2/4] WORKDIR /home/jovyan/work                                 0.0s\n",
      "\u001b[0m\u001b[34m => [3/4] COPY . /home/jovyan/work                                         0.1s\n",
      "\u001b[0m => [4/4] RUN pip install pandas matplotlib                                3.0s\n",
      "\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[2m => => # WARNING: Retrying (Retry(total=2, connect=None, read=None, redirect=No\n",
      "\u001b[0m\u001b[2m => => # ne, status=None)) after connection broken by 'NewConnectionError('<pip\n",
      "\u001b[0m\u001b[2m => => # ._vendor.urllib3.connection.HTTPSConnection object at 0x7f4508445450>:\n",
      "\u001b[0m\u001b[2m => => #  Failed to establish a new connection: [Errno -2] Name or service not \n",
      "\u001b[0m\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[?25h\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[0G\u001b[?25l[+] Building 3.4s (7/8)                                                         \n",
      "\u001b[34m => [internal] load build definition from Dockerfile                       0.0s\n",
      "\u001b[0m\u001b[34m => => transferring dockerfile: 37B                                        0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load .dockerignore                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 2B                                            0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load metadata for docker.io/jupyter/base-notebook:latest    0.0s\n",
      "\u001b[0m\u001b[34m => [1/4] FROM docker.io/jupyter/base-notebook:latest                      0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load build context                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 558.53kB                                      0.0s\n",
      "\u001b[0m\u001b[34m => CACHED [2/4] WORKDIR /home/jovyan/work                                 0.0s\n",
      "\u001b[0m\u001b[34m => [3/4] COPY . /home/jovyan/work                                         0.1s\n",
      "\u001b[0m => [4/4] RUN pip install pandas matplotlib                                3.2s\n",
      "\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[2m => => # WARNING: Retrying (Retry(total=2, connect=None, read=None, redirect=No\n",
      "\u001b[0m\u001b[2m => => # ne, status=None)) after connection broken by 'NewConnectionError('<pip\n",
      "\u001b[0m\u001b[2m => => # ._vendor.urllib3.connection.HTTPSConnection object at 0x7f4508445450>:\n",
      "\u001b[0m\u001b[2m => => #  Failed to establish a new connection: [Errno -2] Name or service not \n",
      "\u001b[0m\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[?25h\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[0G\u001b[?25l[+] Building 3.6s (7/8)                                                         \n",
      "\u001b[34m => [internal] load build definition from Dockerfile                       0.0s\n",
      "\u001b[0m\u001b[34m => => transferring dockerfile: 37B                                        0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load .dockerignore                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 2B                                            0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load metadata for docker.io/jupyter/base-notebook:latest    0.0s\n",
      "\u001b[0m\u001b[34m => [1/4] FROM docker.io/jupyter/base-notebook:latest                      0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load build context                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 558.53kB                                      0.0s\n",
      "\u001b[0m\u001b[34m => CACHED [2/4] WORKDIR /home/jovyan/work                                 0.0s\n",
      "\u001b[0m\u001b[34m => [3/4] COPY . /home/jovyan/work                                         0.1s\n",
      "\u001b[0m => [4/4] RUN pip install pandas matplotlib                                3.3s\n",
      "\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[2m => => # WARNING: Retrying (Retry(total=2, connect=None, read=None, redirect=No\n",
      "\u001b[0m\u001b[2m => => # ne, status=None)) after connection broken by 'NewConnectionError('<pip\n",
      "\u001b[0m\u001b[2m => => # ._vendor.urllib3.connection.HTTPSConnection object at 0x7f4508445450>:\n",
      "\u001b[0m\u001b[2m => => #  Failed to establish a new connection: [Errno -2] Name or service not \n",
      "\u001b[0m\u001b[2m => => # known')': /simple/pandas/                                             \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0m\u001b[?25h\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[0G\u001b[?25l[+] Building 3.7s (7/8)                                                         \n",
      "\u001b[34m => [internal] load build definition from Dockerfile                       0.0s\n",
      "\u001b[0m\u001b[34m => => transferring dockerfile: 37B                                        0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load .dockerignore                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 2B                                            0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load metadata for docker.io/jupyter/base-notebook:latest    0.0s\n",
      "\u001b[0m\u001b[34m => [1/4] FROM docker.io/jupyter/base-notebook:latest                      0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load build context                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 558.53kB                                      0.0s\n",
      "\u001b[0m\u001b[34m => CACHED [2/4] WORKDIR /home/jovyan/work                                 0.0s\n",
      "\u001b[0m\u001b[34m => [3/4] COPY . /home/jovyan/work                                         0.1s\n",
      "\u001b[0m => [4/4] RUN pip install pandas matplotlib                                3.5s\n",
      "\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[2m => => # WARNING: Retrying (Retry(total=2, connect=None, read=None, redirect=No\n",
      "\u001b[0m\u001b[2m => => # ne, status=None)) after connection broken by 'NewConnectionError('<pip\n",
      "\u001b[0m\u001b[2m => => # ._vendor.urllib3.connection.HTTPSConnection object at 0x7f4508445450>:\n",
      "\u001b[0m\u001b[2m => => #  Failed to establish a new connection: [Errno -2] Name or service not \n",
      "\u001b[0m\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[?25h\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[0G\u001b[?25l[+] Building 3.9s (7/8)                                                         \n",
      "\u001b[34m => [internal] load build definition from Dockerfile                       0.0s\n",
      "\u001b[0m\u001b[34m => => transferring dockerfile: 37B                                        0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load .dockerignore                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 2B                                            0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load metadata for docker.io/jupyter/base-notebook:latest    0.0s\n",
      "\u001b[0m\u001b[34m => [1/4] FROM docker.io/jupyter/base-notebook:latest                      0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load build context                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 558.53kB                                      0.0s\n",
      "\u001b[0m\u001b[34m => CACHED [2/4] WORKDIR /home/jovyan/work                                 0.0s\n",
      "\u001b[0m\u001b[34m => [3/4] COPY . /home/jovyan/work                                         0.1s\n",
      "\u001b[0m => [4/4] RUN pip install pandas matplotlib                                3.6s\n",
      "\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[2m => => # WARNING: Retrying (Retry(total=2, connect=None, read=None, redirect=No\n",
      "\u001b[0m\u001b[2m => => # ne, status=None)) after connection broken by 'NewConnectionError('<pip\n",
      "\u001b[0m\u001b[2m => => # ._vendor.urllib3.connection.HTTPSConnection object at 0x7f4508445450>:\n",
      "\u001b[0m\u001b[2m => => #  Failed to establish a new connection: [Errno -2] Name or service not \n",
      "\u001b[0m\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[?25h\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[0G\u001b[?25l[+] Building 4.0s (7/8)                                                         \n",
      "\u001b[34m => [internal] load build definition from Dockerfile                       0.0s\n",
      "\u001b[0m\u001b[34m => => transferring dockerfile: 37B                                        0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load .dockerignore                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 2B                                            0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load metadata for docker.io/jupyter/base-notebook:latest    0.0s\n",
      "\u001b[0m\u001b[34m => [1/4] FROM docker.io/jupyter/base-notebook:latest                      0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load build context                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 558.53kB                                      0.0s\n",
      "\u001b[0m\u001b[34m => CACHED [2/4] WORKDIR /home/jovyan/work                                 0.0s\n",
      "\u001b[0m\u001b[34m => [3/4] COPY . /home/jovyan/work                                         0.1s\n",
      "\u001b[0m => [4/4] RUN pip install pandas matplotlib                                3.8s\n",
      "\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[2m => => # WARNING: Retrying (Retry(total=2, connect=None, read=None, redirect=No\n",
      "\u001b[0m\u001b[2m => => # ne, status=None)) after connection broken by 'NewConnectionError('<pip\n",
      "\u001b[0m\u001b[2m => => # ._vendor.urllib3.connection.HTTPSConnection object at 0x7f4508445450>:\n",
      "\u001b[0m\u001b[2m => => #  Failed to establish a new connection: [Errno -2] Name or service not \n",
      "\u001b[0m\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[?25h\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[0G\u001b[?25l[+] Building 4.2s (7/8)                                                         \n",
      "\u001b[34m => [internal] load build definition from Dockerfile                       0.0s\n",
      "\u001b[0m\u001b[34m => => transferring dockerfile: 37B                                        0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load .dockerignore                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 2B                                            0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load metadata for docker.io/jupyter/base-notebook:latest    0.0s\n",
      "\u001b[0m\u001b[34m => [1/4] FROM docker.io/jupyter/base-notebook:latest                      0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load build context                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 558.53kB                                      0.0s\n",
      "\u001b[0m\u001b[34m => CACHED [2/4] WORKDIR /home/jovyan/work                                 0.0s\n",
      "\u001b[0m\u001b[34m => [3/4] COPY . /home/jovyan/work                                         0.1s\n",
      "\u001b[0m => [4/4] RUN pip install pandas matplotlib                                3.9s\n",
      "\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[2m => => # WARNING: Retrying (Retry(total=2, connect=None, read=None, redirect=No\n",
      "\u001b[0m\u001b[2m => => # ne, status=None)) after connection broken by 'NewConnectionError('<pip\n",
      "\u001b[0m\u001b[2m => => # ._vendor.urllib3.connection.HTTPSConnection object at 0x7f4508445450>:\n",
      "\u001b[0m\u001b[2m => => #  Failed to establish a new connection: [Errno -2] Name or service not \n",
      "\u001b[0m\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[?25h\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[0G\u001b[?25l[+] Building 4.3s (7/8)                                                         \n",
      "\u001b[34m => [internal] load build definition from Dockerfile                       0.0s\n",
      "\u001b[0m\u001b[34m => => transferring dockerfile: 37B                                        0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load .dockerignore                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 2B                                            0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load metadata for docker.io/jupyter/base-notebook:latest    0.0s\n",
      "\u001b[0m\u001b[34m => [1/4] FROM docker.io/jupyter/base-notebook:latest                      0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load build context                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 558.53kB                                      0.0s\n",
      "\u001b[0m\u001b[34m => CACHED [2/4] WORKDIR /home/jovyan/work                                 0.0s\n",
      "\u001b[0m\u001b[34m => [3/4] COPY . /home/jovyan/work                                         0.1s\n",
      "\u001b[0m => [4/4] RUN pip install pandas matplotlib                                4.1s\n",
      "\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[2m => => # WARNING: Retrying (Retry(total=2, connect=None, read=None, redirect=No\n",
      "\u001b[0m\u001b[2m => => # ne, status=None)) after connection broken by 'NewConnectionError('<pip\n",
      "\u001b[0m\u001b[2m => => # ._vendor.urllib3.connection.HTTPSConnection object at 0x7f4508445450>:\n",
      "\u001b[0m\u001b[2m => => #  Failed to establish a new connection: [Errno -2] Name or service not \n",
      "\u001b[0m\u001b[2m => => # known')': /simple/pandas/                                             \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0m\u001b[?25h\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[0G\u001b[?25l[+] Building 4.5s (7/8)                                                         \n",
      "\u001b[34m => [internal] load build definition from Dockerfile                       0.0s\n",
      "\u001b[0m\u001b[34m => => transferring dockerfile: 37B                                        0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load .dockerignore                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 2B                                            0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load metadata for docker.io/jupyter/base-notebook:latest    0.0s\n",
      "\u001b[0m\u001b[34m => [1/4] FROM docker.io/jupyter/base-notebook:latest                      0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load build context                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 558.53kB                                      0.0s\n",
      "\u001b[0m\u001b[34m => CACHED [2/4] WORKDIR /home/jovyan/work                                 0.0s\n",
      "\u001b[0m\u001b[34m => [3/4] COPY . /home/jovyan/work                                         0.1s\n",
      "\u001b[0m => [4/4] RUN pip install pandas matplotlib                                4.2s\n",
      "\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[2m => => # WARNING: Retrying (Retry(total=2, connect=None, read=None, redirect=No\n",
      "\u001b[0m\u001b[2m => => # ne, status=None)) after connection broken by 'NewConnectionError('<pip\n",
      "\u001b[0m\u001b[2m => => # ._vendor.urllib3.connection.HTTPSConnection object at 0x7f4508445450>:\n",
      "\u001b[0m\u001b[2m => => #  Failed to establish a new connection: [Errno -2] Name or service not \n",
      "\u001b[0m\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[?25h\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[0G\u001b[?25l[+] Building 4.6s (7/8)                                                         \n",
      "\u001b[34m => [internal] load build definition from Dockerfile                       0.0s\n",
      "\u001b[0m\u001b[34m => => transferring dockerfile: 37B                                        0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load .dockerignore                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 2B                                            0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load metadata for docker.io/jupyter/base-notebook:latest    0.0s\n",
      "\u001b[0m\u001b[34m => [1/4] FROM docker.io/jupyter/base-notebook:latest                      0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load build context                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 558.53kB                                      0.0s\n",
      "\u001b[0m\u001b[34m => CACHED [2/4] WORKDIR /home/jovyan/work                                 0.0s\n",
      "\u001b[0m\u001b[34m => [3/4] COPY . /home/jovyan/work                                         0.1s\n",
      "\u001b[0m => [4/4] RUN pip install pandas matplotlib                                4.4s\n",
      "\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[2m => => # WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=No\n",
      "\u001b[0m\u001b[2m => => # ne, status=None)) after connection broken by 'NewConnectionError('<pip\n",
      "\u001b[0m\u001b[2m => => # ._vendor.urllib3.connection.HTTPSConnection object at 0x7f4508445d90>:\n",
      "\u001b[0m\u001b[2m => => #  Failed to establish a new connection: [Errno -2] Name or service not \n",
      "\u001b[0m\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[?25h\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[0G\u001b[?25l[+] Building 4.8s (7/8)                                                         \n",
      "\u001b[34m => [internal] load build definition from Dockerfile                       0.0s\n",
      "\u001b[0m\u001b[34m => => transferring dockerfile: 37B                                        0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load .dockerignore                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 2B                                            0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load metadata for docker.io/jupyter/base-notebook:latest    0.0s\n",
      "\u001b[0m\u001b[34m => [1/4] FROM docker.io/jupyter/base-notebook:latest                      0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load build context                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 558.53kB                                      0.0s\n",
      "\u001b[0m\u001b[34m => CACHED [2/4] WORKDIR /home/jovyan/work                                 0.0s\n",
      "\u001b[0m\u001b[34m => [3/4] COPY . /home/jovyan/work                                         0.1s\n",
      "\u001b[0m => [4/4] RUN pip install pandas matplotlib                                4.5s\n",
      "\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[2m => => # WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=No\n",
      "\u001b[0m\u001b[2m => => # ne, status=None)) after connection broken by 'NewConnectionError('<pip\n",
      "\u001b[0m\u001b[2m => => # ._vendor.urllib3.connection.HTTPSConnection object at 0x7f4508445d90>:\n",
      "\u001b[0m\u001b[2m => => #  Failed to establish a new connection: [Errno -2] Name or service not \n",
      "\u001b[0m\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[?25h\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[0G\u001b[?25l[+] Building 4.9s (7/8)                                                         \n",
      "\u001b[34m => [internal] load build definition from Dockerfile                       0.0s\n",
      "\u001b[0m\u001b[34m => => transferring dockerfile: 37B                                        0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load .dockerignore                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 2B                                            0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load metadata for docker.io/jupyter/base-notebook:latest    0.0s\n",
      "\u001b[0m\u001b[34m => [1/4] FROM docker.io/jupyter/base-notebook:latest                      0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load build context                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 558.53kB                                      0.0s\n",
      "\u001b[0m\u001b[34m => CACHED [2/4] WORKDIR /home/jovyan/work                                 0.0s\n",
      "\u001b[0m\u001b[34m => [3/4] COPY . /home/jovyan/work                                         0.1s\n",
      "\u001b[0m => [4/4] RUN pip install pandas matplotlib                                4.7s\n",
      "\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[2m => => # WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=No\n",
      "\u001b[0m\u001b[2m => => # ne, status=None)) after connection broken by 'NewConnectionError('<pip\n",
      "\u001b[0m\u001b[2m => => # ._vendor.urllib3.connection.HTTPSConnection object at 0x7f4508445d90>:\n",
      "\u001b[0m\u001b[2m => => #  Failed to establish a new connection: [Errno -2] Name or service not \n",
      "\u001b[0m\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[?25h\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[0G\u001b[?25l[+] Building 5.1s (7/8)                                                         \n",
      "\u001b[34m => [internal] load build definition from Dockerfile                       0.0s\n",
      "\u001b[0m\u001b[34m => => transferring dockerfile: 37B                                        0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load .dockerignore                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 2B                                            0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load metadata for docker.io/jupyter/base-notebook:latest    0.0s\n",
      "\u001b[0m\u001b[34m => [1/4] FROM docker.io/jupyter/base-notebook:latest                      0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load build context                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 558.53kB                                      0.0s\n",
      "\u001b[0m\u001b[34m => CACHED [2/4] WORKDIR /home/jovyan/work                                 0.0s\n",
      "\u001b[0m\u001b[34m => [3/4] COPY . /home/jovyan/work                                         0.1s\n",
      "\u001b[0m => [4/4] RUN pip install pandas matplotlib                                4.8s\n",
      "\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[2m => => # WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=No\n",
      "\u001b[0m\u001b[2m => => # ne, status=None)) after connection broken by 'NewConnectionError('<pip\n",
      "\u001b[0m\u001b[2m => => # ._vendor.urllib3.connection.HTTPSConnection object at 0x7f4508445d90>:\n",
      "\u001b[0m\u001b[2m => => #  Failed to establish a new connection: [Errno -2] Name or service not \n",
      "\u001b[0m\u001b[2m => => # known')': /simple/pandas/                                             \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0m\u001b[?25h\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[0G\u001b[?25l[+] Building 5.2s (7/8)                                                         \n",
      "\u001b[34m => [internal] load build definition from Dockerfile                       0.0s\n",
      "\u001b[0m\u001b[34m => => transferring dockerfile: 37B                                        0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load .dockerignore                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 2B                                            0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load metadata for docker.io/jupyter/base-notebook:latest    0.0s\n",
      "\u001b[0m\u001b[34m => [1/4] FROM docker.io/jupyter/base-notebook:latest                      0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load build context                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 558.53kB                                      0.0s\n",
      "\u001b[0m\u001b[34m => CACHED [2/4] WORKDIR /home/jovyan/work                                 0.0s\n",
      "\u001b[0m\u001b[34m => [3/4] COPY . /home/jovyan/work                                         0.1s\n",
      "\u001b[0m => [4/4] RUN pip install pandas matplotlib                                5.0s\n",
      "\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[2m => => # WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=No\n",
      "\u001b[0m\u001b[2m => => # ne, status=None)) after connection broken by 'NewConnectionError('<pip\n",
      "\u001b[0m\u001b[2m => => # ._vendor.urllib3.connection.HTTPSConnection object at 0x7f4508445d90>:\n",
      "\u001b[0m\u001b[2m => => #  Failed to establish a new connection: [Errno -2] Name or service not \n",
      "\u001b[0m\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[?25h\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[0G\u001b[?25l[+] Building 5.4s (7/8)                                                         \n",
      "\u001b[34m => [internal] load build definition from Dockerfile                       0.0s\n",
      "\u001b[0m\u001b[34m => => transferring dockerfile: 37B                                        0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load .dockerignore                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 2B                                            0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load metadata for docker.io/jupyter/base-notebook:latest    0.0s\n",
      "\u001b[0m\u001b[34m => [1/4] FROM docker.io/jupyter/base-notebook:latest                      0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load build context                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 558.53kB                                      0.0s\n",
      "\u001b[0m\u001b[34m => CACHED [2/4] WORKDIR /home/jovyan/work                                 0.0s\n",
      "\u001b[0m\u001b[34m => [3/4] COPY . /home/jovyan/work                                         0.1s\n",
      "\u001b[0m => [4/4] RUN pip install pandas matplotlib                                5.1s\n",
      "\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[2m => => # WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=No\n",
      "\u001b[0m\u001b[2m => => # ne, status=None)) after connection broken by 'NewConnectionError('<pip\n",
      "\u001b[0m\u001b[2m => => # ._vendor.urllib3.connection.HTTPSConnection object at 0x7f4508445d90>:\n",
      "\u001b[0m\u001b[2m => => #  Failed to establish a new connection: [Errno -2] Name or service not \n",
      "\u001b[0m\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[?25h\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[0G\u001b[?25l[+] Building 5.5s (7/8)                                                         \n",
      "\u001b[34m => [internal] load build definition from Dockerfile                       0.0s\n",
      "\u001b[0m\u001b[34m => => transferring dockerfile: 37B                                        0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load .dockerignore                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 2B                                            0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load metadata for docker.io/jupyter/base-notebook:latest    0.0s\n",
      "\u001b[0m\u001b[34m => [1/4] FROM docker.io/jupyter/base-notebook:latest                      0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load build context                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 558.53kB                                      0.0s\n",
      "\u001b[0m\u001b[34m => CACHED [2/4] WORKDIR /home/jovyan/work                                 0.0s\n",
      "\u001b[0m\u001b[34m => [3/4] COPY . /home/jovyan/work                                         0.1s\n",
      "\u001b[0m => [4/4] RUN pip install pandas matplotlib                                5.3s\n",
      "\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[2m => => # WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=No\n",
      "\u001b[0m\u001b[2m => => # ne, status=None)) after connection broken by 'NewConnectionError('<pip\n",
      "\u001b[0m\u001b[2m => => # ._vendor.urllib3.connection.HTTPSConnection object at 0x7f4508445d90>:\n",
      "\u001b[0m\u001b[2m => => #  Failed to establish a new connection: [Errno -2] Name or service not \n",
      "\u001b[0m\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[?25h\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[0G\u001b[?25l[+] Building 5.7s (7/8)                                                         \n",
      "\u001b[34m => [internal] load build definition from Dockerfile                       0.0s\n",
      "\u001b[0m\u001b[34m => => transferring dockerfile: 37B                                        0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load .dockerignore                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 2B                                            0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load metadata for docker.io/jupyter/base-notebook:latest    0.0s\n",
      "\u001b[0m\u001b[34m => [1/4] FROM docker.io/jupyter/base-notebook:latest                      0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load build context                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 558.53kB                                      0.0s\n",
      "\u001b[0m\u001b[34m => CACHED [2/4] WORKDIR /home/jovyan/work                                 0.0s\n",
      "\u001b[0m\u001b[34m => [3/4] COPY . /home/jovyan/work                                         0.1s\n",
      "\u001b[0m => [4/4] RUN pip install pandas matplotlib                                5.4s\n",
      "\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[2m => => # WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=No\n",
      "\u001b[0m\u001b[2m => => # ne, status=None)) after connection broken by 'NewConnectionError('<pip\n",
      "\u001b[0m\u001b[2m => => # ._vendor.urllib3.connection.HTTPSConnection object at 0x7f4508445d90>:\n",
      "\u001b[0m\u001b[2m => => #  Failed to establish a new connection: [Errno -2] Name or service not \n",
      "\u001b[0m\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[?25h\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[0G\u001b[?25l[+] Building 5.8s (7/8)                                                         \n",
      "\u001b[34m => [internal] load build definition from Dockerfile                       0.0s\n",
      "\u001b[0m\u001b[34m => => transferring dockerfile: 37B                                        0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load .dockerignore                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 2B                                            0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load metadata for docker.io/jupyter/base-notebook:latest    0.0s\n",
      "\u001b[0m\u001b[34m => [1/4] FROM docker.io/jupyter/base-notebook:latest                      0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load build context                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 558.53kB                                      0.0s\n",
      "\u001b[0m\u001b[34m => CACHED [2/4] WORKDIR /home/jovyan/work                                 0.0s\n",
      "\u001b[0m\u001b[34m => [3/4] COPY . /home/jovyan/work                                         0.1s\n",
      "\u001b[0m => [4/4] RUN pip install pandas matplotlib                                5.6s\n",
      "\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[2m => => # WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=No\n",
      "\u001b[0m\u001b[2m => => # ne, status=None)) after connection broken by 'NewConnectionError('<pip\n",
      "\u001b[0m\u001b[2m => => # ._vendor.urllib3.connection.HTTPSConnection object at 0x7f4508445d90>:\n",
      "\u001b[0m\u001b[2m => => #  Failed to establish a new connection: [Errno -2] Name or service not \n",
      "\u001b[0m\u001b[2m => => # known')': /simple/pandas/                                             \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0m\u001b[?25h\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[0G\u001b[?25l[+] Building 6.0s (7/8)                                                         \n",
      "\u001b[34m => [internal] load build definition from Dockerfile                       0.0s\n",
      "\u001b[0m\u001b[34m => => transferring dockerfile: 37B                                        0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load .dockerignore                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 2B                                            0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load metadata for docker.io/jupyter/base-notebook:latest    0.0s\n",
      "\u001b[0m\u001b[34m => [1/4] FROM docker.io/jupyter/base-notebook:latest                      0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load build context                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 558.53kB                                      0.0s\n",
      "\u001b[0m\u001b[34m => CACHED [2/4] WORKDIR /home/jovyan/work                                 0.0s\n",
      "\u001b[0m\u001b[34m => [3/4] COPY . /home/jovyan/work                                         0.1s\n",
      "\u001b[0m => [4/4] RUN pip install pandas matplotlib                                5.7s\n",
      "\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[2m => => # WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=No\n",
      "\u001b[0m\u001b[2m => => # ne, status=None)) after connection broken by 'NewConnectionError('<pip\n",
      "\u001b[0m\u001b[2m => => # ._vendor.urllib3.connection.HTTPSConnection object at 0x7f4508445d90>:\n",
      "\u001b[0m\u001b[2m => => #  Failed to establish a new connection: [Errno -2] Name or service not \n",
      "\u001b[0m\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[?25h\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[0G\u001b[?25l[+] Building 6.1s (7/8)                                                         \n",
      "\u001b[34m => [internal] load build definition from Dockerfile                       0.0s\n",
      "\u001b[0m\u001b[34m => => transferring dockerfile: 37B                                        0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load .dockerignore                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 2B                                            0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load metadata for docker.io/jupyter/base-notebook:latest    0.0s\n",
      "\u001b[0m\u001b[34m => [1/4] FROM docker.io/jupyter/base-notebook:latest                      0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load build context                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 558.53kB                                      0.0s\n",
      "\u001b[0m\u001b[34m => CACHED [2/4] WORKDIR /home/jovyan/work                                 0.0s\n",
      "\u001b[0m\u001b[34m => [3/4] COPY . /home/jovyan/work                                         0.1s\n",
      "\u001b[0m => [4/4] RUN pip install pandas matplotlib                                5.9s\n",
      "\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[2m => => # WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=No\n",
      "\u001b[0m\u001b[2m => => # ne, status=None)) after connection broken by 'NewConnectionError('<pip\n",
      "\u001b[0m\u001b[2m => => # ._vendor.urllib3.connection.HTTPSConnection object at 0x7f4508445d90>:\n",
      "\u001b[0m\u001b[2m => => #  Failed to establish a new connection: [Errno -2] Name or service not \n",
      "\u001b[0m\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[?25h\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[0G\u001b[?25l[+] Building 6.3s (7/8)                                                         \n",
      "\u001b[34m => [internal] load build definition from Dockerfile                       0.0s\n",
      "\u001b[0m\u001b[34m => => transferring dockerfile: 37B                                        0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load .dockerignore                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 2B                                            0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load metadata for docker.io/jupyter/base-notebook:latest    0.0s\n",
      "\u001b[0m\u001b[34m => [1/4] FROM docker.io/jupyter/base-notebook:latest                      0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load build context                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 558.53kB                                      0.0s\n",
      "\u001b[0m\u001b[34m => CACHED [2/4] WORKDIR /home/jovyan/work                                 0.0s\n",
      "\u001b[0m\u001b[34m => [3/4] COPY . /home/jovyan/work                                         0.1s\n",
      "\u001b[0m => [4/4] RUN pip install pandas matplotlib                                6.0s\n",
      "\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[2m => => # WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=No\n",
      "\u001b[0m\u001b[2m => => # ne, status=None)) after connection broken by 'NewConnectionError('<pip\n",
      "\u001b[0m\u001b[2m => => # ._vendor.urllib3.connection.HTTPSConnection object at 0x7f4508445d90>:\n",
      "\u001b[0m\u001b[2m => => #  Failed to establish a new connection: [Errno -2] Name or service not \n",
      "\u001b[0m\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[?25h\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[0G\u001b[?25l[+] Building 6.4s (7/8)                                                         \n",
      "\u001b[34m => [internal] load build definition from Dockerfile                       0.0s\n",
      "\u001b[0m\u001b[34m => => transferring dockerfile: 37B                                        0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load .dockerignore                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 2B                                            0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load metadata for docker.io/jupyter/base-notebook:latest    0.0s\n",
      "\u001b[0m\u001b[34m => [1/4] FROM docker.io/jupyter/base-notebook:latest                      0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load build context                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 558.53kB                                      0.0s\n",
      "\u001b[0m\u001b[34m => CACHED [2/4] WORKDIR /home/jovyan/work                                 0.0s\n",
      "\u001b[0m\u001b[34m => [3/4] COPY . /home/jovyan/work                                         0.1s\n",
      "\u001b[0m => [4/4] RUN pip install pandas matplotlib                                6.2s\n",
      "\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[2m => => # WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=No\n",
      "\u001b[0m\u001b[2m => => # ne, status=None)) after connection broken by 'NewConnectionError('<pip\n",
      "\u001b[0m\u001b[2m => => # ._vendor.urllib3.connection.HTTPSConnection object at 0x7f4508445d90>:\n",
      "\u001b[0m\u001b[2m => => #  Failed to establish a new connection: [Errno -2] Name or service not \n",
      "\u001b[0m\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[?25h\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[0G\u001b[?25l[+] Building 6.6s (7/8)                                                         \n",
      "\u001b[34m => [internal] load build definition from Dockerfile                       0.0s\n",
      "\u001b[0m\u001b[34m => => transferring dockerfile: 37B                                        0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load .dockerignore                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 2B                                            0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load metadata for docker.io/jupyter/base-notebook:latest    0.0s\n",
      "\u001b[0m\u001b[34m => [1/4] FROM docker.io/jupyter/base-notebook:latest                      0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load build context                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 558.53kB                                      0.0s\n",
      "\u001b[0m\u001b[34m => CACHED [2/4] WORKDIR /home/jovyan/work                                 0.0s\n",
      "\u001b[0m\u001b[34m => [3/4] COPY . /home/jovyan/work                                         0.1s\n",
      "\u001b[0m => [4/4] RUN pip install pandas matplotlib                                6.3s\n",
      "\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[2m => => # WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=No\n",
      "\u001b[0m\u001b[2m => => # ne, status=None)) after connection broken by 'NewConnectionError('<pip\n",
      "\u001b[0m\u001b[2m => => # ._vendor.urllib3.connection.HTTPSConnection object at 0x7f4508445d90>:\n",
      "\u001b[0m\u001b[2m => => #  Failed to establish a new connection: [Errno -2] Name or service not \n",
      "\u001b[0m\u001b[2m => => # known')': /simple/pandas/                                             \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0m\u001b[?25h\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[0G\u001b[?25l[+] Building 6.7s (7/8)                                                         \n",
      "\u001b[34m => [internal] load build definition from Dockerfile                       0.0s\n",
      "\u001b[0m\u001b[34m => => transferring dockerfile: 37B                                        0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load .dockerignore                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 2B                                            0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load metadata for docker.io/jupyter/base-notebook:latest    0.0s\n",
      "\u001b[0m\u001b[34m => [1/4] FROM docker.io/jupyter/base-notebook:latest                      0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load build context                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 558.53kB                                      0.0s\n",
      "\u001b[0m\u001b[34m => CACHED [2/4] WORKDIR /home/jovyan/work                                 0.0s\n",
      "\u001b[0m\u001b[34m => [3/4] COPY . /home/jovyan/work                                         0.1s\n",
      "\u001b[0m => [4/4] RUN pip install pandas matplotlib                                6.5s\n",
      "\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[2m => => # WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=No\n",
      "\u001b[0m\u001b[2m => => # ne, status=None)) after connection broken by 'NewConnectionError('<pip\n",
      "\u001b[0m\u001b[2m => => # ._vendor.urllib3.connection.HTTPSConnection object at 0x7f4508445d90>:\n",
      "\u001b[0m\u001b[2m => => #  Failed to establish a new connection: [Errno -2] Name or service not \n",
      "\u001b[0m\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[?25h\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[0G\u001b[?25l[+] Building 6.9s (7/8)                                                         \n",
      "\u001b[34m => [internal] load build definition from Dockerfile                       0.0s\n",
      "\u001b[0m\u001b[34m => => transferring dockerfile: 37B                                        0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load .dockerignore                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 2B                                            0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load metadata for docker.io/jupyter/base-notebook:latest    0.0s\n",
      "\u001b[0m\u001b[34m => [1/4] FROM docker.io/jupyter/base-notebook:latest                      0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load build context                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 558.53kB                                      0.0s\n",
      "\u001b[0m\u001b[34m => CACHED [2/4] WORKDIR /home/jovyan/work                                 0.0s\n",
      "\u001b[0m\u001b[34m => [3/4] COPY . /home/jovyan/work                                         0.1s\n",
      "\u001b[0m => [4/4] RUN pip install pandas matplotlib                                6.6s\n",
      "\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[2m => => # WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=No\n",
      "\u001b[0m\u001b[2m => => # ne, status=None)) after connection broken by 'NewConnectionError('<pip\n",
      "\u001b[0m\u001b[2m => => # ._vendor.urllib3.connection.HTTPSConnection object at 0x7f4508445d90>:\n",
      "\u001b[0m\u001b[2m => => #  Failed to establish a new connection: [Errno -2] Name or service not \n",
      "\u001b[0m\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[?25h\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[0G\u001b[?25l[+] Building 7.0s (7/8)                                                         \n",
      "\u001b[34m => [internal] load build definition from Dockerfile                       0.0s\n",
      "\u001b[0m\u001b[34m => => transferring dockerfile: 37B                                        0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load .dockerignore                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 2B                                            0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load metadata for docker.io/jupyter/base-notebook:latest    0.0s\n",
      "\u001b[0m\u001b[34m => [1/4] FROM docker.io/jupyter/base-notebook:latest                      0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load build context                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 558.53kB                                      0.0s\n",
      "\u001b[0m\u001b[34m => CACHED [2/4] WORKDIR /home/jovyan/work                                 0.0s\n",
      "\u001b[0m\u001b[34m => [3/4] COPY . /home/jovyan/work                                         0.1s\n",
      "\u001b[0m => [4/4] RUN pip install pandas matplotlib                                6.8s\n",
      "\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[2m => => # WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=No\n",
      "\u001b[0m\u001b[2m => => # ne, status=None)) after connection broken by 'NewConnectionError('<pip\n",
      "\u001b[0m\u001b[2m => => # ._vendor.urllib3.connection.HTTPSConnection object at 0x7f4508445d90>:\n",
      "\u001b[0m\u001b[2m => => #  Failed to establish a new connection: [Errno -2] Name or service not \n",
      "\u001b[0m\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[?25h\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[0G\u001b[?25l[+] Building 7.2s (7/8)                                                         \n",
      "\u001b[34m => [internal] load build definition from Dockerfile                       0.0s\n",
      "\u001b[0m\u001b[34m => => transferring dockerfile: 37B                                        0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load .dockerignore                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 2B                                            0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load metadata for docker.io/jupyter/base-notebook:latest    0.0s\n",
      "\u001b[0m\u001b[34m => [1/4] FROM docker.io/jupyter/base-notebook:latest                      0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load build context                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 558.53kB                                      0.0s\n",
      "\u001b[0m\u001b[34m => CACHED [2/4] WORKDIR /home/jovyan/work                                 0.0s\n",
      "\u001b[0m\u001b[34m => [3/4] COPY . /home/jovyan/work                                         0.1s\n",
      "\u001b[0m => [4/4] RUN pip install pandas matplotlib                                6.9s\n",
      "\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[2m => => # WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=No\n",
      "\u001b[0m\u001b[2m => => # ne, status=None)) after connection broken by 'NewConnectionError('<pip\n",
      "\u001b[0m\u001b[2m => => # ._vendor.urllib3.connection.HTTPSConnection object at 0x7f4508445d90>:\n",
      "\u001b[0m\u001b[2m => => #  Failed to establish a new connection: [Errno -2] Name or service not \n",
      "\u001b[0m\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[?25h\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[0G\u001b[?25l[+] Building 7.3s (7/8)                                                         \n",
      "\u001b[34m => [internal] load build definition from Dockerfile                       0.0s\n",
      "\u001b[0m\u001b[34m => => transferring dockerfile: 37B                                        0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load .dockerignore                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 2B                                            0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load metadata for docker.io/jupyter/base-notebook:latest    0.0s\n",
      "\u001b[0m\u001b[34m => [1/4] FROM docker.io/jupyter/base-notebook:latest                      0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load build context                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 558.53kB                                      0.0s\n",
      "\u001b[0m\u001b[34m => CACHED [2/4] WORKDIR /home/jovyan/work                                 0.0s\n",
      "\u001b[0m\u001b[34m => [3/4] COPY . /home/jovyan/work                                         0.1s\n",
      "\u001b[0m => [4/4] RUN pip install pandas matplotlib                                7.1s\n",
      "\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[2m => => # WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=No\n",
      "\u001b[0m\u001b[2m => => # ne, status=None)) after connection broken by 'NewConnectionError('<pip\n",
      "\u001b[0m\u001b[2m => => # ._vendor.urllib3.connection.HTTPSConnection object at 0x7f4508445d90>:\n",
      "\u001b[0m\u001b[2m => => #  Failed to establish a new connection: [Errno -2] Name or service not \n",
      "\u001b[0m\u001b[2m => => # known')': /simple/pandas/                                             \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0m\u001b[?25h\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[0G\u001b[?25l[+] Building 7.5s (7/8)                                                         \n",
      "\u001b[34m => [internal] load build definition from Dockerfile                       0.0s\n",
      "\u001b[0m\u001b[34m => => transferring dockerfile: 37B                                        0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load .dockerignore                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 2B                                            0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load metadata for docker.io/jupyter/base-notebook:latest    0.0s\n",
      "\u001b[0m\u001b[34m => [1/4] FROM docker.io/jupyter/base-notebook:latest                      0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load build context                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 558.53kB                                      0.0s\n",
      "\u001b[0m\u001b[34m => CACHED [2/4] WORKDIR /home/jovyan/work                                 0.0s\n",
      "\u001b[0m\u001b[34m => [3/4] COPY . /home/jovyan/work                                         0.1s\n",
      "\u001b[0m => [4/4] RUN pip install pandas matplotlib                                7.2s\n",
      "\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[2m => => # WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=No\n",
      "\u001b[0m\u001b[2m => => # ne, status=None)) after connection broken by 'NewConnectionError('<pip\n",
      "\u001b[0m\u001b[2m => => # ._vendor.urllib3.connection.HTTPSConnection object at 0x7f4508445d90>:\n",
      "\u001b[0m\u001b[2m => => #  Failed to establish a new connection: [Errno -2] Name or service not \n",
      "\u001b[0m\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[?25h\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[0G\u001b[?25l[+] Building 7.6s (7/8)                                                         \n",
      "\u001b[34m => [internal] load build definition from Dockerfile                       0.0s\n",
      "\u001b[0m\u001b[34m => => transferring dockerfile: 37B                                        0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load .dockerignore                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 2B                                            0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load metadata for docker.io/jupyter/base-notebook:latest    0.0s\n",
      "\u001b[0m\u001b[34m => [1/4] FROM docker.io/jupyter/base-notebook:latest                      0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load build context                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 558.53kB                                      0.0s\n",
      "\u001b[0m\u001b[34m => CACHED [2/4] WORKDIR /home/jovyan/work                                 0.0s\n",
      "\u001b[0m\u001b[34m => [3/4] COPY . /home/jovyan/work                                         0.1s\n",
      "\u001b[0m => [4/4] RUN pip install pandas matplotlib                                7.4s\n",
      "\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[2m => => # WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=No\n",
      "\u001b[0m\u001b[2m => => # ne, status=None)) after connection broken by 'NewConnectionError('<pip\n",
      "\u001b[0m\u001b[2m => => # ._vendor.urllib3.connection.HTTPSConnection object at 0x7f4508445d90>:\n",
      "\u001b[0m\u001b[2m => => #  Failed to establish a new connection: [Errno -2] Name or service not \n",
      "\u001b[0m\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[?25h\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[0G\u001b[?25l[+] Building 7.8s (7/8)                                                         \n",
      "\u001b[34m => [internal] load build definition from Dockerfile                       0.0s\n",
      "\u001b[0m\u001b[34m => => transferring dockerfile: 37B                                        0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load .dockerignore                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 2B                                            0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load metadata for docker.io/jupyter/base-notebook:latest    0.0s\n",
      "\u001b[0m\u001b[34m => [1/4] FROM docker.io/jupyter/base-notebook:latest                      0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load build context                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 558.53kB                                      0.0s\n",
      "\u001b[0m\u001b[34m => CACHED [2/4] WORKDIR /home/jovyan/work                                 0.0s\n",
      "\u001b[0m\u001b[34m => [3/4] COPY . /home/jovyan/work                                         0.1s\n",
      "\u001b[0m => [4/4] RUN pip install pandas matplotlib                                7.5s\n",
      "\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[2m => => # WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=No\n",
      "\u001b[0m\u001b[2m => => # ne, status=None)) after connection broken by 'NewConnectionError('<pip\n",
      "\u001b[0m\u001b[2m => => # ._vendor.urllib3.connection.HTTPSConnection object at 0x7f4508445d90>:\n",
      "\u001b[0m\u001b[2m => => #  Failed to establish a new connection: [Errno -2] Name or service not \n",
      "\u001b[0m\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[?25h\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[0G\u001b[?25l[+] Building 7.9s (7/8)                                                         \n",
      "\u001b[34m => [internal] load build definition from Dockerfile                       0.0s\n",
      "\u001b[0m\u001b[34m => => transferring dockerfile: 37B                                        0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load .dockerignore                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 2B                                            0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load metadata for docker.io/jupyter/base-notebook:latest    0.0s\n",
      "\u001b[0m\u001b[34m => [1/4] FROM docker.io/jupyter/base-notebook:latest                      0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load build context                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 558.53kB                                      0.0s\n",
      "\u001b[0m\u001b[34m => CACHED [2/4] WORKDIR /home/jovyan/work                                 0.0s\n",
      "\u001b[0m\u001b[34m => [3/4] COPY . /home/jovyan/work                                         0.1s\n",
      "\u001b[0m => [4/4] RUN pip install pandas matplotlib                                7.7s\n",
      "\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[2m => => # WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=No\n",
      "\u001b[0m\u001b[2m => => # ne, status=None)) after connection broken by 'NewConnectionError('<pip\n",
      "\u001b[0m\u001b[2m => => # ._vendor.urllib3.connection.HTTPSConnection object at 0x7f4508445d90>:\n",
      "\u001b[0m\u001b[2m => => #  Failed to establish a new connection: [Errno -2] Name or service not \n",
      "\u001b[0m\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[?25h\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[0G\u001b[?25l[+] Building 8.1s (7/8)                                                         \n",
      "\u001b[34m => [internal] load build definition from Dockerfile                       0.0s\n",
      "\u001b[0m\u001b[34m => => transferring dockerfile: 37B                                        0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load .dockerignore                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 2B                                            0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load metadata for docker.io/jupyter/base-notebook:latest    0.0s\n",
      "\u001b[0m\u001b[34m => [1/4] FROM docker.io/jupyter/base-notebook:latest                      0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load build context                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 558.53kB                                      0.0s\n",
      "\u001b[0m\u001b[34m => CACHED [2/4] WORKDIR /home/jovyan/work                                 0.0s\n",
      "\u001b[0m\u001b[34m => [3/4] COPY . /home/jovyan/work                                         0.1s\n",
      "\u001b[0m => [4/4] RUN pip install pandas matplotlib                                7.8s\n",
      "\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[2m => => # WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=No\n",
      "\u001b[0m\u001b[2m => => # ne, status=None)) after connection broken by 'NewConnectionError('<pip\n",
      "\u001b[0m\u001b[2m => => # ._vendor.urllib3.connection.HTTPSConnection object at 0x7f4508445d90>:\n",
      "\u001b[0m\u001b[2m => => #  Failed to establish a new connection: [Errno -2] Name or service not \n",
      "\u001b[0m\u001b[2m => => # known')': /simple/pandas/                                             \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0m\u001b[?25h\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[0G\u001b[?25l[+] Building 8.2s (7/8)                                                         \n",
      "\u001b[34m => [internal] load build definition from Dockerfile                       0.0s\n",
      "\u001b[0m\u001b[34m => => transferring dockerfile: 37B                                        0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load .dockerignore                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 2B                                            0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load metadata for docker.io/jupyter/base-notebook:latest    0.0s\n",
      "\u001b[0m\u001b[34m => [1/4] FROM docker.io/jupyter/base-notebook:latest                      0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load build context                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 558.53kB                                      0.0s\n",
      "\u001b[0m\u001b[34m => CACHED [2/4] WORKDIR /home/jovyan/work                                 0.0s\n",
      "\u001b[0m\u001b[34m => [3/4] COPY . /home/jovyan/work                                         0.1s\n",
      "\u001b[0m => [4/4] RUN pip install pandas matplotlib                                8.0s\n",
      "\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[2m => => # WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=No\n",
      "\u001b[0m\u001b[2m => => # ne, status=None)) after connection broken by 'NewConnectionError('<pip\n",
      "\u001b[0m\u001b[2m => => # ._vendor.urllib3.connection.HTTPSConnection object at 0x7f4508445d90>:\n",
      "\u001b[0m\u001b[2m => => #  Failed to establish a new connection: [Errno -2] Name or service not \n",
      "\u001b[0m\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[?25h\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[0G\u001b[?25l[+] Building 8.4s (7/8)                                                         \n",
      "\u001b[34m => [internal] load build definition from Dockerfile                       0.0s\n",
      "\u001b[0m\u001b[34m => => transferring dockerfile: 37B                                        0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load .dockerignore                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 2B                                            0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load metadata for docker.io/jupyter/base-notebook:latest    0.0s\n",
      "\u001b[0m\u001b[34m => [1/4] FROM docker.io/jupyter/base-notebook:latest                      0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load build context                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 558.53kB                                      0.0s\n",
      "\u001b[0m\u001b[34m => CACHED [2/4] WORKDIR /home/jovyan/work                                 0.0s\n",
      "\u001b[0m\u001b[34m => [3/4] COPY . /home/jovyan/work                                         0.1s\n",
      "\u001b[0m => [4/4] RUN pip install pandas matplotlib                                8.1s\n",
      "\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[2m => => # WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=No\n",
      "\u001b[0m\u001b[2m => => # ne, status=None)) after connection broken by 'NewConnectionError('<pip\n",
      "\u001b[0m\u001b[2m => => # ._vendor.urllib3.connection.HTTPSConnection object at 0x7f4508445d90>:\n",
      "\u001b[0m\u001b[2m => => #  Failed to establish a new connection: [Errno -2] Name or service not \n",
      "\u001b[0m\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[?25h\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[0G\u001b[?25l[+] Building 8.5s (7/8)                                                         \n",
      "\u001b[34m => [internal] load build definition from Dockerfile                       0.0s\n",
      "\u001b[0m\u001b[34m => => transferring dockerfile: 37B                                        0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load .dockerignore                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 2B                                            0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load metadata for docker.io/jupyter/base-notebook:latest    0.0s\n",
      "\u001b[0m\u001b[34m => [1/4] FROM docker.io/jupyter/base-notebook:latest                      0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load build context                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 558.53kB                                      0.0s\n",
      "\u001b[0m\u001b[34m => CACHED [2/4] WORKDIR /home/jovyan/work                                 0.0s\n",
      "\u001b[0m\u001b[34m => [3/4] COPY . /home/jovyan/work                                         0.1s\n",
      "\u001b[0m => [4/4] RUN pip install pandas matplotlib                                8.3s\n",
      "\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[2m => => # WARNING: Retrying (Retry(total=0, connect=None, read=None, redirect=No\n",
      "\u001b[0m\u001b[2m => => # ne, status=None)) after connection broken by 'NewConnectionError('<pip\n",
      "\u001b[0m\u001b[2m => => # ._vendor.urllib3.connection.HTTPSConnection object at 0x7f4508446790>:\n",
      "\u001b[0m\u001b[2m => => #  Failed to establish a new connection: [Errno -2] Name or service not \n",
      "\u001b[0m\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[?25h\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[0G\u001b[?25l[+] Building 8.7s (7/8)                                                         \n",
      "\u001b[34m => [internal] load build definition from Dockerfile                       0.0s\n",
      "\u001b[0m\u001b[34m => => transferring dockerfile: 37B                                        0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load .dockerignore                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 2B                                            0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load metadata for docker.io/jupyter/base-notebook:latest    0.0s\n",
      "\u001b[0m\u001b[34m => [1/4] FROM docker.io/jupyter/base-notebook:latest                      0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load build context                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 558.53kB                                      0.0s\n",
      "\u001b[0m\u001b[34m => CACHED [2/4] WORKDIR /home/jovyan/work                                 0.0s\n",
      "\u001b[0m\u001b[34m => [3/4] COPY . /home/jovyan/work                                         0.1s\n",
      "\u001b[0m => [4/4] RUN pip install pandas matplotlib                                8.4s\n",
      "\u001b[2m => => # ._vendor.urllib3.connection.HTTPSConnection object at 0x7f4508446790>:\n",
      "\u001b[0m\u001b[2m => => #  Failed to establish a new connection: [Errno -2] Name or service not \n",
      "\u001b[0m\u001b[2m => => # known')': /simple/pandas/                                             \n",
      "\u001b[0m\u001b[2m => => # ERROR: Could not find a version that satisfies the requirement pandas \n",
      "\u001b[0m\u001b[2m => => # (from versions: none)                                                 \n",
      "\u001b[0m\u001b[2m => => # ERROR: No matching distribution found for pandas                      \n",
      "\u001b[0m\u001b[?25h\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[1A\u001b[0G\u001b[?25l[+] Building 8.7s (8/8) FINISHED                                                \n",
      "\u001b[34m => [internal] load build definition from Dockerfile                       0.0s\n",
      "\u001b[0m\u001b[34m => => transferring dockerfile: 37B                                        0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load .dockerignore                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 2B                                            0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load metadata for docker.io/jupyter/base-notebook:latest    0.0s\n",
      "\u001b[0m\u001b[34m => [1/4] FROM docker.io/jupyter/base-notebook:latest                      0.0s\n",
      "\u001b[0m\u001b[34m => [internal] load build context                                          0.0s\n",
      "\u001b[0m\u001b[34m => => transferring context: 558.53kB                                      0.0s\n",
      "\u001b[0m\u001b[34m => CACHED [2/4] WORKDIR /home/jovyan/work                                 0.0s\n",
      "\u001b[0m\u001b[34m => [3/4] COPY . /home/jovyan/work                                         0.1s\n",
      "\u001b[0m\u001b[31m => ERROR [4/4] RUN pip install pandas matplotlib                          8.5s\n",
      "\u001b[0m                                                                                \n",
      "                                                                                \n",
      "                                                                                \n",
      "                                                                                \n",
      "                                                                                \n",
      "                                                                                \n",
      "\u001b[6A\u001b[0G\u001b[?25h------\n",
      " > [4/4] RUN pip install pandas matplotlib:\n",
      "#8 0.754 WARNING: Retrying (Retry(total=4, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7f4508444210>: Failed to establish a new connection: [Errno -2] Name or service not known')': /simple/pandas/\n",
      "#8 1.260 WARNING: Retrying (Retry(total=3, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7f4508444d90>: Failed to establish a new connection: [Errno -2] Name or service not known')': /simple/pandas/\n",
      "#8 2.271 WARNING: Retrying (Retry(total=2, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7f4508445450>: Failed to establish a new connection: [Errno -2] Name or service not known')': /simple/pandas/\n",
      "#8 4.278 WARNING: Retrying (Retry(total=1, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7f4508445d90>: Failed to establish a new connection: [Errno -2] Name or service not known')': /simple/pandas/\n",
      "#8 8.287 WARNING: Retrying (Retry(total=0, connect=None, read=None, redirect=None, status=None)) after connection broken by 'NewConnectionError('<pip._vendor.urllib3.connection.HTTPSConnection object at 0x7f4508446790>: Failed to establish a new connection: [Errno -2] Name or service not known')': /simple/pandas/\n",
      "#8 8.305 ERROR: Could not find a version that satisfies the requirement pandas (from versions: none)\n",
      "#8 8.305 ERROR: No matching distribution found for pandas\n",
      "------\n",
      "executor failed running [/bin/bash -o pipefail -c pip install pandas matplotlib]: exit code: 1\n"
     ]
    }
   ],
   "source": [
    "!docker build -t jupyter_server ."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The build process will create a Docker image with your Jupyter Notebook server configuration.\n",
    "\n",
    "To run the Jupyter Notebook server in a container based on the image you just built, use the following command:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[I 2023-08-12 19:19:48.106 ServerApp] Package notebook took 0.0000s to import\n",
      "[I 2023-08-12 19:19:48.131 ServerApp] Package jupyter_lsp took 0.0172s to import\n",
      "[W 2023-08-12 19:19:48.131 ServerApp] A `_jupyter_server_extension_points` function was not found in jupyter_lsp. Instead, a `_jupyter_server_extension_paths` function was found and will be used for now. This function name will be deprecated in future releases of Jupyter Server.\n",
      "[I 2023-08-12 19:19:48.140 ServerApp] Package jupyter_server_terminals took 0.0086s to import\n",
      "[I 2023-08-12 19:19:48.141 ServerApp] Package jupyterlab took 0.0000s to import\n",
      "[I 2023-08-12 19:19:48.705 ServerApp] Package nbclassic took 0.0019s to import\n",
      "[W 2023-08-12 19:19:48.707 ServerApp] A `_jupyter_server_extension_points` function was not found in nbclassic. Instead, a `_jupyter_server_extension_paths` function was found and will be used for now. This function name will be deprecated in future releases of Jupyter Server.\n",
      "[I 2023-08-12 19:19:48.707 ServerApp] Package notebook_shim took 0.0000s to import\n",
      "[W 2023-08-12 19:19:48.707 ServerApp] A `_jupyter_server_extension_points` function was not found in notebook_shim. Instead, a `_jupyter_server_extension_paths` function was found and will be used for now. This function name will be deprecated in future releases of Jupyter Server.\n",
      "[I 2023-08-12 19:19:48.708 ServerApp] jupyter_lsp | extension was successfully linked.\n",
      "[I 2023-08-12 19:19:48.712 ServerApp] jupyter_server_terminals | extension was successfully linked.\n",
      "[I 2023-08-12 19:19:48.716 ServerApp] jupyterlab | extension was successfully linked.\n",
      "[I 2023-08-12 19:19:48.720 ServerApp] nbclassic | extension was successfully linked.\n",
      "[I 2023-08-12 19:19:48.724 ServerApp] notebook | extension was successfully linked.\n",
      "[I 2023-08-12 19:19:48.725 ServerApp] Writing Jupyter server cookie secret to /home/jovyan/.local/share/jupyter/runtime/jupyter_cookie_secret\n",
      "[I 2023-08-12 19:19:48.943 ServerApp] notebook_shim | extension was successfully linked.\n",
      "[I 2023-08-12 19:19:48.971 ServerApp] notebook_shim | extension was successfully loaded.\n",
      "[I 2023-08-12 19:19:48.973 ServerApp] jupyter_lsp | extension was successfully loaded.\n",
      "[I 2023-08-12 19:19:48.974 ServerApp] jupyter_server_terminals | extension was successfully loaded.\n",
      "[I 2023-08-12 19:19:48.976 LabApp] JupyterLab extension loaded from /opt/conda/lib/python3.11/site-packages/jupyterlab\n",
      "[I 2023-08-12 19:19:48.976 LabApp] JupyterLab application directory is /opt/conda/share/jupyter/lab\n",
      "[I 2023-08-12 19:19:48.977 LabApp] Extension Manager is 'pypi'.\n",
      "[I 2023-08-12 19:19:48.980 ServerApp] jupyterlab | extension was successfully loaded.\n",
      "[I 2023-08-12 19:19:48.983 ServerApp] nbclassic | extension was successfully loaded.\n",
      "[I 2023-08-12 19:19:48.986 ServerApp] notebook | extension was successfully loaded.\n",
      "[I 2023-08-12 19:19:48.986 ServerApp] Serving notebooks from local directory: /home/jovyan/work\n",
      "[I 2023-08-12 19:19:48.987 ServerApp] Jupyter Server 2.7.0 is running at:\n",
      "[I 2023-08-12 19:19:48.987 ServerApp] http://6786521305f5:8889/tree?token=bf10af5a3f4bd15a0d617b26c87563066f6683d495949e4e\n",
      "[I 2023-08-12 19:19:48.987 ServerApp]     http://127.0.0.1:8889/tree?token=bf10af5a3f4bd15a0d617b26c87563066f6683d495949e4e\n",
      "[I 2023-08-12 19:19:48.987 ServerApp] Use Control-C to stop this server and shut down all kernels (twice to skip confirmation).\n",
      "[C 2023-08-12 19:19:48.990 ServerApp] \n",
      "    \n",
      "    To access the server, open this file in a browser:\n",
      "        file:///home/jovyan/.local/share/jupyter/runtime/jpserver-7-open.html\n",
      "    Or copy and paste one of these URLs:\n",
      "        http://6786521305f5:8889/tree?token=bf10af5a3f4bd15a0d617b26c87563066f6683d495949e4e\n",
      "        http://127.0.0.1:8889/tree?token=bf10af5a3f4bd15a0d617b26c87563066f6683d495949e4e\n",
      "[I 2023-08-12 19:19:49.564 ServerApp] Skipped non-installed server(s): bash-language-server, dockerfile-language-server-nodejs, javascript-typescript-langserver, jedi-language-server, julia-language-server, pyright, python-language-server, python-lsp-server, r-languageserver, sql-language-server, texlab, typescript-language-server, unified-language-server, vscode-css-languageserver-bin, vscode-html-languageserver-bin, vscode-json-languageserver-bin, yaml-language-server\n",
      "0.00s - Debugger warning: It seems that frozen modules are being used, which may\n",
      "0.00s - make the debugger miss breakpoints. Please pass -Xfrozen_modules=off\n",
      "0.00s - to python to disable frozen modules.\n",
      "0.00s - Note: Debugging will proceed. Set PYDEVD_DISABLE_FILE_VALIDATION=1 to disable this validation.\n",
      "^C\n",
      "\n",
      "  _   _          _      _\n",
      " | | | |_ __  __| |__ _| |_ ___\n",
      " | |_| | '_ \\/ _` / _` |  _/ -_)\n",
      "  \\___/| .__/\\__,_\\__,_|\\__\\___|\n",
      "       |_|\n",
      "                                                                           \n",
      "Read the migration plan to Notebook 7 to learn about the new features and the actions to take if you are using extensions.\n",
      "\n",
      "https://jupyter-notebook.readthedocs.io/en/latest/migrate_to_notebook7.html\n",
      "\n",
      "Please note that updating to Notebook 7 might break some of your extensions.\n",
      "\n",
      "[I 2023-08-12 19:22:22.457 ServerApp] Interrupted...\n"
     ]
    }
   ],
   "source": [
    "!docker run -p 8889:8889 jupyter_server"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The **-p 8889:8889** maps port 8889 from the container to port 8889 on the host machine. You can then access the Jupyter Notebook server by opening a web browser and navigating to http://localhost:8889 in the host machine. You'll see the Jupyter Notebook interface and can start working with your notebooks."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
